solver/llbsolver/file/backend.go

```diff
@@ -205,38 +205,37 @@ func docopy(ctx context.Context, src, dest string, action pb.FileActionCopy, u *
 		copy.WithXAttrErrorHandler(xattrErrorHandler),
 	}
 
+	var m []string
 	if !action.AllowWildcard {
-		if action.AttemptUnpackDockerCompatibility {
-			if ok, err := unpack(ctx, src, srcPath, dest, destPath, ch, timestampToTime(action.Timestamp)); err != nil {
-				return err
-			} else if ok {
-				return nil
-			}
+		m = append(m, src)
+	} else {
+		var err error
+		m, err = copy.ResolveWildcards(src, srcPath, action.FollowSymlink)
+		if err != nil {
+			return err
 		}
-		return copy.Copy(ctx, src, srcPath, dest, destPath, opt...)
-	}
 
-	m, err := copy.ResolveWildcards(src, srcPath, action.FollowSymlink)
-	if err != nil {
-		return err
-	}
-
-	if len(m) == 0 {
-		if action.AllowEmptyWildcard {
-			return nil
+		if len(m) == 0 {
+			if action.AllowEmptyWildcard {
+				return nil
+			}
+			return errors.Errorf("%s not found", srcPath)
 		}
-		return errors.Errorf("%s not found", srcPath)
 	}
 
 	for _, s := range m {
+		d := destPath
+		if action.Parents {
+			d = filepath.Join("/", d, s)
+		}
 		if action.AttemptUnpackDockerCompatibility {
-			if ok, err := unpack(ctx, src, s, dest, destPath, ch, timestampToTime(action.Timestamp)); err != nil {
+			if ok, err := unpack(ctx, src, s, dest, d, ch, timestampToTime(action.Timestamp)); err != nil {
 				return err
 			} else if ok {
-				continue
+				return nil
```

Comment:
:eyes: I think this line should stay the same, we still want to process the rest of the list.

Reply:
Oh yeah, thanks. Stumbled into it too while implementing tests... :sweat: